File: packages/pulse/swift/Tests/PulseTests/Harness.swift
import XCTest

@testable import Pulse

/// Test harness: a deterministic two-endpoint world with a programmable faulty
/// channel and a virtual clock. Port of the TypeScript `World` harness. No real
/// sockets, no real time — every spec §9 failure is reproducible and fast.
final class World {
    enum Dir { case aToB, bToA }

    var now = 0
    let a: Endpoint
    let b: Endpoint

    var deliveredA: [(seq: UInt64, payload: [UInt8])] = []
    var deliveredB: [(seq: UInt64, payload: [UInt8])] = []
    var resetsA: [(fromSeq: UInt64, peerEpoch: String)] = []
    var resetsB: [(fromSeq: UInt64, peerEpoch: String)] = []

    private var linkUp = false
    private var linkAvailable = false
    private var dropCount: [Dir: Int] = [.aToB: 0, .bToA: 0]
    private var dupCount: [Dir: Int] = [.aToB: 0, .bToA: 0]
    private var blackholed = false
    private var reorderBuffer: [(dir: Dir, bytes: [UInt8])]?

    init(a: Endpoint, b: Endpoint) {
        self.a = a
        self.b = b
    }

    // MARK: - Clock

    func advance(_ ms: Int) {
        let target = now + ms
        while true {
            let da = a.nextDeadline() ?? Int.max
            let db = b.nextDeadline() ?? Int.max
            let next = min(da, db)
            if next == Int.max || next > target { break }
            if next > now { now = next }
            pump(a.onTick(now), .aToB)
            pump(b.onTick(now), .bToA)
        }
        now = target
        pump(a.onTick(now), .aToB)
        pump(b.onTick(now), .bToA)
    }

    // MARK: - Link control

    func connect() {
        linkUp = true
        linkAvailable = true
        blackholed = false
        pump(a.onConnected(now), .aToB)
        pump(b.onConnected(now), .bToA)
    }

    func disconnect() {
        linkUp = false
        linkAvailable = false
        pump(a.onDisconnected(now), .aToB)
        pump(b.onDisconnected(now), .bToA)
    }

    func blackhole() {
        blackholed = true
        linkAvailable = false
    }

    func reopen() { connect() }

    // MARK: - Application send

    @discardableResult
    func sendA(_ payload: [UInt8]) -> UInt64 {
        let (seq, effects) = a.send(payload)
        pump(effects, .aToB)
        return seq
    }

    @discardableResult
    func sendB(_ payload: [UInt8]) -> UInt64 {
        let (seq, effects) = b.send(payload)
        pump(effects, .bToA)
        return seq
    }

    // MARK: - Fault programming

    func dropNext(_ dir: Dir, _ n: Int) { dropCount[dir, default: 0] += n }
    func duplicateNext(_ dir: Dir, _ n: Int) { dupCount[dir, default: 0] += n }

    func beginReorder() { reorderBuffer = [] }
    func flushReordered() {
        let buf = reorderBuffer ?? []
        reorderBuffer = nil
        for f in buf.reversed() { deliver(f) }
    }

    // MARK: - Internal plumbing

    private func pump(_ effects: [Effect], _ dir: Dir) {
        for e in effects { applyEffect(e, dir) }
    }

    private func applyEffect(_ e: Effect, _ dir: Dir) {
        let producedByA = (dir == .aToB)
        switch e {
        case let .transmit(bytes):
            enqueue((dir: dir, bytes: bytes))
        case let .deliver(seq, payload):
            if producedByA { deliveredA.append((seq, payload)) }
            else { deliveredB.append((seq, payload)) }
        case let .resetInbound(fromSeq, peerEpoch):
            if producedByA { resetsA.append((fromSeq, peerEpoch)) }
            else { resetsB.append((fromSeq, peerEpoch)) }
        case .open:
            if linkAvailable {
                connect()
            } else {
                pump(producedByA ? a.onDisconnected(now) : b.onDisconnected(now), dir)
            }
        case .close:
            if linkUp || blackholed { disconnect() }
        }
    }

    private func enqueue(_ f: (dir: Dir, bytes: [UInt8])) {
        if reorderBuffer != nil {
            reorderBuffer!.append(f)
            return
        }
        deliver(f)
    }

    private func deliver(_ f: (dir: Dir, bytes: [UInt8])) {
        if !linkUp || blackholed { return }
        if dropCount[f.dir, default: 0] > 0 {
            dropCount[f.dir]! -= 1
            return
        }
        route(f)
        if dupCount[f.dir, default: 0] > 0 {
            dupCount[f.dir]! -= 1
            route(f)
        }
    }

    private func route(_ f: (dir: Dir, bytes: [UInt8])) {
        if f.dir == .aToB {
            pump(b.onBytes(f.bytes, now), .bToA)
        } else {
            pump(a.onBytes(f.bytes, now), .aToB)
        }
    }
}

// MARK: - helpers

func marker(_ n: Int) -> [UInt8] { [UInt8(n & 0xFF)] }

func payloads(_ d: [(seq: UInt64, payload: [UInt8])]) -> [Int] {
    d.map { Int($0.payload.first ?? 255) }
}

func seqs(_ d: [(seq: UInt64, payload: [UInt8])]) -> [UInt64] {
    d.map { $0.seq }
}

File: packages/pulse/ts/src/__tests__/harness.ts
/**
 * Test harness: a deterministic two-endpoint world with a programmable faulty
 * channel and a virtual clock. No real sockets, no real time — so every
 * real-world failure in spec §9 is reproducible and fast.
 *
 * The harness models the LINK the way TCP/TLS/WebSocket actually behaves:
 * ordered and fail-stop while connected (spec §1). "Drop", "reorder",
 * "duplicate" here are deliberately ADVERSARIAL — they inject conditions the
 * live link normally prevents, so we can prove the endpoint's resume/dedup
 * logic is defensive even if the transport misbehaves.
 */

import { Endpoint } from '../endpoint.js';
import type { Effect, EndpointOptions } from '../types.js';

export type Dir = 'AtoB' | 'BtoA';

/** A frame in flight, tagged with direction and the sender's identity. */
interface InFlight {
  dir: Dir;
  bytes: Uint8Array;
  /** Monotonic send order, for deterministic tie-breaking on equal arrival. */
  order: number;
}

/**
 * Drives two endpoints and the channel between them under one virtual clock.
 *
 * The channel is "connected" or not per direction pair; when connected, bytes
 * flow subject to the installed faults. Faults are opt-in and explicit.
 */
export class World {
  now = 0;
  readonly a: Endpoint;
  readonly b: Endpoint;

  /** Payloads delivered to the application at each side, in delivery order. */
  readonly deliveredA: Array<{ seq: bigint; payload: Uint8Array }> = [];
  readonly deliveredB: Array<{ seq: bigint; payload: Uint8Array }> = [];
  /** reset-inbound effects observed at each side. */
  readonly resetsA: Array<{ fromSeq: bigint; peerEpoch: string }> = [];
  readonly resetsB: Array<{ fromSeq: bigint; peerEpoch: string }> = [];
  /** Highest acked seq observed at each side (the delivery-confirmation floor
   *  for messages that side SENT). Last value = latest. */
  readonly ackedA: bigint[] = [];
  readonly ackedB: bigint[] = [];

  private linkUp = false;
  /** Whether a dial would currently succeed. Distinct from linkUp (currently
   *  connected): after a disconnect/blackhole the path is unavailable until the
   *  test calls reopen(), so auto-dials fail and reschedule. */
  private linkAvailable = false;
  // Fault programming:
  private dropCount: Record<Dir, number> = { AtoB: 0, BtoA: 0 };
  private dropPredicate: ((dir: Dir, bytes: Uint8Array) => boolean) | null = null;
  private dupCount: Record<Dir, number> = { AtoB: 0, BtoA: 0 };
  private blackholed = false;
  private reorderBuffer: InFlight[] | null = null;
  /** When true, the underlying path is healthy: an endpoint's own `open`
   *  (auto-reconnect) succeeds, and a liveness `close` becomes a transient
   *  disconnect the endpoint recovers from by itself. Used to model "the
   *  network is now good; let pulse converge" without the test manually
   *  re-dialing after every internal teardown. */
  private healing = false;
  /** One-way propagation delay in virtual ms. 0 ⇒ synchronous delivery (the
   *  default; keeps existing scenarios' timing exact). */
  private latencyMs = 0;
  /** Deterministic per-frame jitter added on top of latencyMs: frame k is
   *  delayed by latencyMs + (k % (jitterMs+1)). 0 ⇒ no jitter. */
  private jitterMs = 0;
  private frameCounter = 0;
  /** Frames propagating through the wire, ascending by arrival time. */
  private inFlight: Array<{ f: InFlight; arriveAt: number }> = [];

  constructor(aOpts: EndpointOptions, bOpts: EndpointOptions) {
    this.a = new Endpoint(aOpts);
    this.b = new Endpoint(bOpts);
  }

  // ── Deterministic random source usable by endpoints (0.5 unless overridden)

  // ── Clock ──────────────────────────────────────────────────────────────

  /** Advance virtual time by `ms`, ticking both endpoints at each of their
   *  requested deadlines AND delivering in-flight frames at their arrival
   *  time (and at the final instant). */
  advance(ms: number): void {
    const target = this.now + ms;
    // Loop: repeatedly jump to the earliest pending event (a timer deadline or
    // an in-flight frame arrival) within the window.
    for (;;) {
      const da = this.a.nextDeadline();
      const db = this.b.nextDeadline();
      const dw = this.earliestArrival();
      const next = Math.min(
        da ?? Number.POSITIVE_INFINITY,
        db ?? Number.POSITIVE_INFINITY,
        dw ?? Number.POSITIVE_INFINITY,
      );
      if (next === Number.POSITIVE_INFINITY || next > target) break;
      if (next > this.now) this.now = next;
      this.deliverArrivalsUpTo(this.now);
      this.pump(this.a.onTick(this.now), 'AtoB');
      this.pump(this.b.onTick(this.now), 'BtoA');
    }
    this.now = target;
    this.deliverArrivalsUpTo(this.now);
    this.pump(this.a.onTick(this.now), 'AtoB');
    this.pump(this.b.onTick(this.now), 'BtoA');
  }

  private earliestArrival(): number | null {
    let min: number | null = null;
    for (const e of this.inFlight) {
      if (min === null || e.arriveAt < min) min = e.arriveAt;
    }
    return min;
  }

  /** Deliver every in-flight frame whose arrival time has come. Frames whose
   *  link died mid-flight are dropped (fail-stop): a real socket loses bytes
   *  still in its buffer when it closes. */
  private deliverArrivalsUpTo(t: number): void {
    if (this.inFlight.length === 0) return;
    // Ascending by arrival, then by send order (frameCounter) for determinism.
    const ready = this.inFlight
      .filter((e) => e.arriveAt <= t)
      .sort((x, y) => x.arriveAt - y.arriveAt || x.f.order - y.f.order);
    this.inFlight = this.inFlight.filter((e) => e.arriveAt > t);
    for (const e of ready) this.route(e.f);
  }

  // ── Link control ─────────────────────────────────────────────────────────

  connect(): void {
    this.linkUp = true;
    this.linkAvailable = true;
    this.blackholed = false;
    this.pump(this.a.onConnected(this.now), 'AtoB');
    this.pump(this.b.onConnected(this.now), 'BtoA');
  }

  /** Graceful disconnect: both sides get onDisconnected. Frames still in
   *  flight are lost (fail-stop: a closing socket drops buffered bytes). */
  disconnect(): void {
    this.linkUp = false;
    if (!this.healing) this.linkAvailable = false;
    this.inFlight = [];
    this.pump(this.a.onDisconnected(this.now), 'AtoB');
    this.pump(this.b.onDisconnected(this.now), 'BtoA');
  }

  /** Put the world into a healthy state and let the endpoints reconnect and
   *  converge on their own: the path is available, so each endpoint's own
   *  `open` (auto-reconnect) succeeds, and any liveness teardown self-heals.
   *  Models "the network is now good" — the settle phase of a fault program. */
  heal(): void {
    this.healing = true;
    this.blackholed = false;
    this.linkAvailable = true;
    this.latencyMs = 0;
    this.jitterMs = 0;
    this.dropCount = { AtoB: 0, BtoA: 0 };
    this.dropPredicate = null;
    this.dupCount = { AtoB: 0, BtoA: 0 };
    // If the link was blackholed, the endpoints still THINK they're connected
    // but no traffic flows; force a clean reconnect so liveness + resume run.
    if (this.linkUp) this.disconnect();
    this.connect();
  }

  /** Half-open: the wire silently black-holes; NO onDisconnected fires.
   *  The endpoints still believe they are connected until liveness trips.
   *  In-flight bytes vanish (the black hole eats them). */
  blackhole(): void {
    this.blackholed = true;
    this.linkAvailable = false;
    this.inFlight = [];
  }

  /** Reconnect after the endpoints have (independently) decided to reopen.
   *  Mirrors an adapter acting on an `open` effect. */
  reopen(): void {
    this.connect();
  }

  // ── Application send ───────────────────────────────────────────────────────

  sendA(payload: Uint8Array): bigint {
    const { seq, effects } = this.a.send(payload);
    this.pump(effects, 'AtoB');
    return seq;
  }
  sendB(payload: Uint8Array): bigint {
    const { seq, effects } = this.b.send(payload);
    this.pump(effects, 'BtoA');
    return seq;
  }

  // ── Fault programming ──────────────────────────────────────────────────────

  dropNext(dir: Dir, n: number): void {
    this.dropCount[dir] += n;
  }
  dropMatching(pred: (dir: Dir, bytes: Uint8Array) => boolean): void {
    this.dropPredicate = pred;
  }
  clearDropMatching(): void {
    this.dropPredicate = null;
  }
  duplicateNext(dir: Dir, n: number): void {
    this.dupCount[dir] += n;
  }

  /** Set one-way propagation delay (virtual ms) for all subsequent frames.
   *  Frames already in flight keep their scheduled arrival. */
  latency(ms: number): void {
    this.latencyMs = ms;
  }
  /** Add deterministic per-frame jitter on top of latency: frame k gets an
   *  extra `k % (ms+1)` delay, so consecutive frames spread out (and can even
   *  cross — exercising the endpoint's in-order guarantee under real jitter). */
  jitter(ms: number): void {
    this.jitterMs = ms;
  }

  /** Begin buffering all frames instead of delivering; released, reversed,
   *  by {@link flushReordered}. Simulates adversarial reordering. */
  beginReorder(): void {
    this.reorderBuffer = [];
  }
  flushReordered(): void {
    const buf = this.reorderBuffer ?? [];
    this.reorderBuffer = null;
    // deliver in REVERSE order to maximally stress ordering logic
    for (let i = buf.length - 1; i >= 0; i--) {
      const f = buf[i]!;
      this.deliver(f);
    }
  }

  // ── Internal plumbing ──────────────────────────────────────────────────────

  /** Carry out a batch of effects from the endpoint identified by outbound dir. */
  private pump(effects: Effect[], dir: Dir): void {
    for (const e of effects) this.applyEffect(e, dir);
  }

  private applyEffect(e: Effect, dir: Dir): void {
    // `dir` is the OUTBOUND direction of the endpoint that produced this effect
    // (A ⇒ 'AtoB', B ⇒ 'BtoA'). `transmit` crosses to the other side; `deliver`
    // and `reset-inbound` belong to the PRODUCING side's own application.
    const producedByA = dir === 'AtoB';
    switch (e.t) {
      case 'transmit':
        this.enqueue({ dir, bytes: e.bytes, order: ++this.frameCounter });
        break;
      case 'deliver':
        (producedByA ? this.deliveredA : this.deliveredB).push({
          seq: e.seq,
          payload: e.payload,
        });
        break;
      case 'reset-inbound':
        (producedByA ? this.resetsA : this.resetsB).push({
          fromSeq: e.fromSeq,
          peerEpoch: e.peerEpoch,
        });
        break;
      case 'acked':
        (producedByA ? this.ackedA : this.ackedB).push(e.seqUpTo);
        break;
      case 'open':
        // Endpoint asked to dial. The two endpoints share one link, so a dial
        // from either side brings it up for both — but only once (idempotent):
        // a second open while already up is a no-op, not a re-connect that would
        // reset the peer's liveness state.
        if (this.linkUp) break;
        if (this.linkAvailable) {
          this.connect();
        } else {
          // Model a FAILED dial by feeding onDisconnected back, which reschedules
          // the next retry — proving the endpoint keeps trying and never wedges.
          this.pump(
            producedByA ? this.a.onDisconnected(this.now) : this.b.onDisconnected(this.now),
            dir,
          );
        }
        break;
      case 'close':
        // Endpoint declared its link dead (e.g. liveness). Model the socket
        // teardown: that side (and, for a real duplex socket, the other) get
        // onDisconnected. We only disconnect if we currently believe up.
        if (this.linkUp || this.blackholed) this.disconnect();
        break;
    }
  }

  private enqueue(f: InFlight): void {
    if (this.reorderBuffer) {
      this.reorderBuffer.push(f);
      return;
    }
    this.deliver(f);
  }

  private deliver(f: InFlight): void {
    // Link down or black-holed ⇒ bytes vanish (fail-stop / silent).
    if (!this.linkUp || this.blackholed) return;

    // Drop-by-predicate:
    if (this.dropPredicate && this.dropPredicate(f.dir, f.bytes)) return;
    // Drop-next-N:
    if (this.dropCount[f.dir] > 0) {
      this.dropCount[f.dir] -= 1;
      return;
    }

    this.propagate(f);

    // Duplicate-next-N: a second copy enters the wire (same latency path).
    if (this.dupCount[f.dir] > 0) {
      this.dupCount[f.dir] -= 1;
      this.propagate({ ...f, order: ++this.frameCounter });
    }
  }

  /** Put a frame onto the wire. With zero latency it arrives synchronously
   *  (preserving existing scenarios' exact timing); with latency/jitter it
   *  becomes an in-flight frame delivered later by the clock. */
  private propagate(f: InFlight): void {
    if (this.latencyMs === 0 && this.jitterMs === 0) {
      this.route(f);
      return;
    }
    const extra = this.jitterMs === 0 ? 0 : f.order % (this.jitterMs + 1);
    this.inFlight.push({ f, arriveAt: this.now + this.latencyMs + extra });
  }

  private route(f: InFlight): void {
    if (f.dir === 'AtoB') {
      this.pump(this.b.onBytes(f.bytes, this.now), 'BtoA');
    } else {
      this.pump(this.a.onBytes(f.bytes, this.now), 'AtoB');
    }
  }
}

// ── Small helpers for tests ──────────────────────────────────────────────────

let epochCounter = 0;
/** Deterministic unique epoch per call (tests need distinct cold-start epochs). */
export function freshEpoch(prefix = 'ep'): string {
  epochCounter += 1;
  return `${prefix}-${epochCounter}`;
}

export function bytes(...vals: number[]): Uint8Array {
  return new Uint8Array(vals);
}

/** Payloads as ascending byte markers, for identity assertions. */
export function marker(n: number): Uint8Array {
  return new Uint8Array([n & 0xff]);
}

export function payloadsOf(
  delivered: Array<{ seq: bigint; payload: Uint8Array }>,
): number[] {
  return delivered.map((d) => d.payload[0] ?? -1);
}

export function seqsOf(
  delivered: Array<{ seq: bigint; payload: Uint8Array }>,
): bigint[] {
  return delivered.map((d) => d.seq);
}
